src/analyzer.c

```diff
@@ -535,23 +535,7 @@ int AddNode(GwNode *nd, char *aliasname)
 /* add multiple nodes (if array) */
 int AddNodeUnroll(GwNode *nd, char *aliasname)
 {
-#ifdef WAVE_ARRAY_SUPPORT
-    if (nd->array_height <= 1)
-#endif
-    {
-        return (AddNodeTraceReturn(nd, aliasname, NULL));
-    }
-#ifdef WAVE_ARRAY_SUPPORT
-    else {
-        unsigned int i;
-        int rc = 1;
-
-        for (i = 0; i < nd->array_height; i++) {
-            rc |= AddNodeTraceReturn(nd + i, aliasname, NULL);
-        }
-        return (rc);
-    }
-#endif
+    return (AddNodeTraceReturn(nd, aliasname, NULL));
```

Comment:
`AddNodeUnroll` was only used for the array support. We could simplify the code a bit by removing it entirely and by replacing 'AddNodeUnroll` with `AddNodeTraceReturn` everywhere it is used.

Reply:
Thanks for your advice.

I just noticed that `AddNode` and `AddNodeUnroll` are now identical.
For now, I prefer replacing `AddNodeUnroll` with `AddNode` since its name is more straightforward and consistent with other `Add` functions in `analyzer.c` (`AddNode`, `AddVector`).

However, I'm still trying to figure out the relationship between `node` and `trace`, and what they stand for. These two data structure come from the CVS era, so I can't find much information why they are added.

Also, does the "return" in `AddNodeTraceReturn` mean "Add a node, **then** return the trace (via `GwTrace **tret`)"?

Reply:
Node is what the loaders will generate.
Trace wraps more data around that to make the node information useful to be viewed onscreen.
IIRC, BitVector is generally used also for bit extraction/concatenation from a Node.  In a Trace, there is a union that will contain either a Node pointer or a BitVector pointer.

Reply:
I also had some issues understanding how this part of the code worked at first. `AddTrace` does heavily rely on the global state in `GLOBALS`, which makes code like the savefile loader hard to understand.

I haven't spend a lot of time to clean up that part of the code, because I'm planning to replace the current `GwTrace`  (aka `TraceEnt`, `Trptr` in LTS) with a new GObject based type. I've already done some work on this and implemented a new parser for the `gtkw` savefile format, but that hasn't been merged yet.